### PR TITLE
Prefer instancetype over id in initializers

### PR DESCRIPTION
**Pull request by CodaFi**
- Uses `instancetype` over `id` where possible

### CODE REVIEW

File: build-mac/mailcore2.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1732,7 +1732,7 @@
 		C63CD69016BE566E00DB18F1 /* MCHTMLCleaner.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = MCHTMLCleaner.h; sourceTree = "<group>"; };
 		C63D315A17C9155C00A4D993 /* MCIMAPIdentity.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = MCIMAPIdentity.cpp; sourceTree = "<group>"; };
 		C63D315B17C9155C00A4D993 /* MCIMAPIdentity.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = MCIMAPIdentity.h; sourceTree = "<group>"; };
-		C63D316017C92D8300A4D993 /* MCOIMAPIdentity.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = MCOIMAPIdentity.h; sourceTree = "<group>"; };
```
Comment (dinhviethoa) [inline build-mac/mailcore2.xcodeproj/project.pbxproj]: Do you know why these changes happened?

File: src/objc/abstract/MCOAbstractMessage.h [hunk 2]
```diff
@@ -40,19 +41,19 @@ namespace mailcore {
  It will return an array of MCOIMAPPart for MCOIMAPMessage.
  It will return an array of MCOAttachment for MCOMessageParser.
  It will return an array of MCOAttachment for MCOMessageBuilder. */
-- (NSArray * /* MCOAbstractPart */) attachments;
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOAbstractMessage.h]: I actually preferred methods since it's not a no-cost accessor.

File: src/objc/abstract/MCOAbstractPart.h [hunk 3]
```diff
@@ -105,14 +105,14 @@ namespace mailcore {
 - (NSString *) contentTypeParameterValueForName:(NSString *)name;
 
 /** Returns an array with the names of all content type parameters.*/
-- (NSArray * /* NSString */) allContentTypeParametersNames;
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOAbstractPart.h]: Same here, don't change it to property.

File: src/objc/abstract/MCOMessageHeader.h [hunk 4]
```diff
@@ -76,25 +76,25 @@
...
-- (MCOMessageHeader *) replyHeaderWithExcludedRecipients:(NSArray * /* MCOAddress */)excludedRecipients;
+- (MCOMessageHeader *) replyHeaderWithExcludedRecipients:(NSArray<MCOAddress *> *)excludedRecipients;
 
 /** Returns a header that can be used as a base for a reply all message.*/
-- (MCOMessageHeader *) replyAllHeaderWithExcludedRecipients:(NSArray * /* MCOAddress */)excludedRecipients;
+- (MCOMessageHeader *) replyAllHeaderWithExcludedRecipients:(NSArray<MCOAddress *> *)excludedRecipients;
 
 /** Returns a header that can be used as a base for a forward message.*/
-- (MCOMessageHeader *) forwardHeader;
+@property (nonatomic, copy, readonly) MCOMessageHeader * forwardHeader;
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOMessageHeader.h]: same here.

File: src/objc/abstract/MCOMessageHeader.h [hunk 5]
```diff
@@ -76,25 +76,25 @@
 - (NSString *) extraHeaderValueForName:(NSString *)name;
 
 /** Returns an array with the names of all custom headers.*/
-- (NSArray * /* NSString */) allExtraHeadersNames;
+@property (nonatomic, copy, readonly) NSArray<NSString *> * allExtraHeadersNames;
```
Comment (dinhviethoa) [inline src/objc/abstract/MCOMessageHeader.h]: same here and below.

File: src/objc/imap/MCOIMAPFetchNamespaceOperation.h [hunk 6]
```diff
@@ -33,7 +35,7 @@ NS_ASSUME_NONNULL_BEGIN
  - On failure, `error` will be set with `MCOErrorDomain` as domain and an 
    error code available in `MCOConstants.h`, `namespaces` will be nil
 */
-- (void) start:(void (^)(NSError * __nullable error, NSDictionary * __nullable namespaces))completionBlock;
+- (void) start:(void (^)(NSError * __nullable error, NSDictionary<NSString *, MCOIMAPNamespace *> * _Nullable namespaces))completionBlock;
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPFetchNamespaceOperation.h]: what's the difference between __nullable and _Nullable?

File: src/objc/imap/MCOIMAPFolder.h [hunk 7]
```diff
@@ -15,10 +15,11 @@
 #import <Foundation/Foundation.h>
 #import <MailCore/MCOConstants.h>
 
+NS_ASSUME_NONNULL_BEGIN
 @interface MCOIMAPFolder : NSObject <NSCopying>
 
 /** The folder's path, like for example INBOX.Archive */
-@property (nonatomic, copy) NSString * path;
+@property (nullable, nonatomic, copy) NSString * path;
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPFolder.h]: I'm skeptical about this one.

File: src/objc/imap/MCOIMAPPart.h [hunk 8]
```diff
@@ -16,10 +16,11 @@
 
 /** Represents a single IMAP message part */
 
+NS_ASSUME_NONNULL_BEGIN
 @interface MCOIMAPPart : MCOAbstractPart <NSCoding>
 
 /** A part identifier looks like 1.2.1 */
-@property (nonatomic, copy) NSString * partID;
+@property (nullable, nonatomic, copy) NSString * partID;
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPPart.h]: please remove nullable.

File: src/objc/imap/MCOIMAPPart.h [hunk 9]
```diff
@@ -31,8 +32,9 @@
  Returns the decoded size of the part.
  For example, for a part that's encoded with base64, it will return actual_size * 3/4.
 */
-- (unsigned int) decodedSize;
+@property (nonatomic) unsigned int decodedSize;
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPPart.h]: move back to method.

File: src/objc/imap/MCOIMAPSession.mm [hunk 10]
```diff
@@ -249,13 +249,13 @@ - (MCOIMAPAppendMessageOperation *)appendMessageOperationWithFolder:(NSString *)
                                                         messageData:(NSData *)messageData
                                                               flags:(MCOMessageFlag)flags
 {
-    return [self appendMessageOperationWithFolder:folder messageData:messageData flags:flags customFlags:NULL];
+    return [self appendMessageOperationWithFolder:folder messageData:messageData flags:flags customFlags:@[]];
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPSession.mm]: Why did you need to change nil to @[]?

**Comment by CodaFi** [comment]
Re The property changes: I made them because in Swift all of these would be computed properties.  They're all on objects we own and control the semantics for, and are only for value objects that we are free to modify.  Plus they allow us to easily add and remove nullability annotations in the future to these classes.

File: src/objc/imap/MCOIMAPSession.mm [hunk 11]
```diff
@@ -475,14 +475,14 @@ - (MCOIMAPOperation *) storeFlagsOperationWithFolder:(NSString *)folder
                                                 kind:(MCOIMAPStoreFlagsRequestKind)kind
                                                flags:(MCOMessageFlag)flags
 {
-    return [self storeFlagsOperationWithFolder:folder uids:uids kind:kind flags:flags customFlags:NULL];
+    return [self storeFlagsOperationWithFolder:folder uids:uids kind:kind flags:flags customFlags:@[]];
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPSession.mm]: same here.

File: src/objc/imap/MCOIMAPSession.mm (hunk 10 above)
Comment (CodaFi) [inline src/objc/imap/MCOIMAPSession.mm]: That parameter is now `nonnull`

File: src/objc/imap/MCOIMAPFolder.h (hunk 7 above)
Comment (CodaFi) [inline src/objc/imap/MCOIMAPFolder.h]: I was too, but I can't find any evidence to support nonnull

File: src/objc/imap/MCOIMAPPart.h (hunk 8 above)
Comment (CodaFi) [inline src/objc/imap/MCOIMAPPart.h]: I can't find evidence to support nonnull here

**Comment by dinhviethoa** [comment]
> Re The property changes: I made them because in Swift all of these would be computed properties. They're all on objects we own and control the semantics for, and are only for value objects that we are free to modify. Plus they allow us to easily add and remove nullability annotations in the future to these classes.

Please don't disrupt Obj-C semantic with Swift :)

File: src/objc/imap/MCOIMAPSession.mm (hunk 10 above)
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPSession.mm]: please allow null.

File: src/objc/imap/MCOIMAPPart.h (hunk 8 above)
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPPart.h]: is it nonnull by default??

**Comment by dinhviethoa** [comment]
Why don't we start with just the "instancetype" change?

**Comment by dinhviethoa** [comment]
The generics could also be a non-disruptive change.

**Comment by dinhviethoa** [comment]
You could submit them as separate pull requests.

**Comment by CodaFi** [comment]
Done.
